### DISCUSSION TITLE
v0.15.0

### DISCUSSION THREAD

**Discussion by martinvonz** [jj-vcs/jj#3238] [General]
### Breaking changes

* The minimum supported Rust version (MSRV) is now 1.76.0.

* The on-disk index format changed. New index files will be created
  automatically, but it can fail if the repository is co-located and predates
  Git GC issues [#815](https://github.com/martinvonz/jj/issues/815). If
  reindexing failed, you'll need to clean up corrupted operation history by
  `jj op abandon ..<bad operation ID>`.

* Dropped support for the "legacy" graph-drawing style. Use "ascii" for a very
  similar result.

* The default log output no longer lists all tagged heads. Set `revsets.log =
  "@ | ancestors(immutable_heads().., 2) | heads(immutable_heads())"` to restore
  the old behavior.

* Dropped support for the deprecated `:` revset operator. Use `::` instead.

* `jj rebase --skip-empty` no longer abandons commits that were already empty
  before the rebase.

### New features

* Added support for commit signing and verification. This comes with 
  out-of-the-box support for the following backends:
  * GnuPG
  * SSH

* Templates now support logical operators: `||`, `&&`, `!`

* Templates now support the `self` keyword, which is the current commit in `jj
  log`/`obslog` templates.

* `jj show` now accepts `-T`/`--template` option to render its output using
  template

* `jj config list` now accepts `-T`/`--template` option.

* `jj git fetch` now accepts `-b` as a shorthand for `--branch`, making it more
  consistent with other commands that accept a branch

* In the templating language, Timestamps now have a `.local()` method for
  converting to the local timezone.

* `jj next/prev` now infer `--edit` when you're already editing a non-head
  commit (a commit with children).

* A new built-in pager named `:builtin` is available on all platforms,
  implemented with [minus](https://github.com/arijit79/minus/)

* Set config `ui.log-synthetic-elided-nodes = true` to make `jj log` include
  synthetic nodes in the graph where some revisions were elided
  ([#1252](https://github.com/martinvonz/jj/issues/1252),
  [#2971](https://github.com/martinvonz/jj/issues/2971)). This may become the
  default depending on feedback.

* When creating a new workspace, the sparse patterns are now copied over from
  the current workspace.

* `jj git init --colocate` can now import an existing Git repository. This is
  equivalent to `jj git init --git-repo=.`.

* `jj git fetch` now automatically prints new remote branches and tags by default.

* `--verbose/-v` is now `--debug` (no short option since it's not intended to be used often)

* `jj move --from/--to` can now be abbreviated to `jj move -f/-t`

* `jj commit`/`diffedit`/`move`/`resolve`/`split`/`squash`/`unsquash` now accept
  `--tool=<NAME>` option to override the default.
   [#2575](https://github.com/martinvonz/jj/issues/2575)

* Added completions for [Nushell](https://nushell.sh) to `jj util completion`

* `jj branch list` now supports a `--tracked/-t` option which can be used to
  show tracked branches only. Omits local Git-tracking branches by default.

* Commands producing diffs now accept a `--context` flag for the number of
  lines of context to show.

* `jj` commands with the `-T`/`--template` option now provide a hint containing
  defined template names when no argument is given, assisting the user in making
  a selection.

### Fixed bugs

* On Windows, symlinks in the repo are now supported when Developer Mode is enabled.
  When symlink support is unavailable, they will be materialized as regular files in the
  working copy (instead of resulting in a crash).
  [#2](https://github.com/martinvonz/jj/issues/2)
  
* On Windows, the `:builtin` pager is now used by default, rather than being
  disabled entirely.

* Auto-rebase now preserves the shape of history even for merge commits where
  one parent is an ancestor of another.
  [#2600](https://github.com/martinvonz/jj/issues/2600)

### Contributors

Thanks to the people who made this release happen!

* Aleksey Kuznetsov (@zummenix)
* Anton Bulakh (@necauqua)
* Anton Älgmyr (@algmyr)
* Austin Seipp (@thoughtpolice)
* Benjamin Brittain (@benbrittain)
* Benjamin Tan (@bnjmnt4n)
* Daehyeok Mun (@daehyeok)
* Daniel Ploch (@torquestomp)
* Evan Mesterhazy (@emesterhazy)
* gulbanana (@gulbanana)
* Ilya Grigoriev (@ilyagr)
* Jonathan Tan (@jonathantanmy)
* Julien Vincent (@julienvincent)
* jyn (@jyn514)
* Martin von Zweigbergk (@martinvonz)
* Paulo Coelho (@prscoelho)
* Philip Metzger (@PhilipMetzger)
* Poliorcetics (@poliorcetics)
* Stephen Jennings (@jennings)
* Vladimir (@0xdeafbeef)
* Yuya Nishihara (@yuja)

<hr /><em>This discussion was created from the release <a href='https://github.com/martinvonz/jj/releases/tag/v0.15.0'>v0.15.0</a>.</em>

**Comment by necauqua** [comment]
Oof, the signing is in (but also without the sign command?. don't remember fully atm), but we haven't merged the verification bit in, and the changelog entry is from when it was a single PR I think 🙃

**Reply by martinvonz** [reply]
Both the text above and the text in the CHANGELOG.md can be changed. What should I change it to?

**Reply by necauqua** [reply]
Something similar to that maybe?. Only half of the signing stuff made it into the release :)

```
Partial support for commit signing.
Currently you can configure jj to "keep" commit signatures by making new ones for rewritten commits,
and to sign new commits when they are created.

Signature verification and an explicit sign command will land in 0.16.

This comes with out-of-the-box support for the following backends:
- GnuPG
- SSH
```

**Reply by julienvincent** [reply]
I think I would avoid stating the rest will land in a prticular upcoming version, but the rest looks good to me.

P.s: Sorry for missing this!

**Comment by amiryal** [comment]
> * Set config `ui.log-synthetic-elided-nodes = true` to make `jj log` include
>       synthetic nodes in the graph where some revisions were elided
>       ([#1252](https://github.com/martinvonz/jj/issues/1252),
>       [#2971](https://github.com/martinvonz/jj/issues/2971)). This may become the
>       default depending on feedback.

I love it! Some feedback:

The current choice of Unicode character (DOTTED CIRCLE) renders as double the width when followed by a space in my terminal (Kitty on Ubuntu Jammy 22.04):
![image](https://github.com/martinvonz/jj/assets/105565/45e5a821-48c2-4cc8-9ca8-679a0b188e2e)

Therefore, I am in support of such exploration as #3124. It is worth noting that it’s not an issue on another workstation running Kitty on Ubuntu Mantic 23.10.

I would also do away with the wording `(elided revisions)`. Simply spacing the graph where elided commits are present would be better, in my taste. I.e. compared to this:
```
$ jj log -r 'z|heads(::)'
@  wzkrkuur (no email set) 2024-03-08 11:32:28.000 +02:00 b6b41359
│  (empty) (no description set)
◌  (elided revisions)
│ ◉  tpstlust support+octocat@github.com 2018-05-10 12:55:19.000 -05:00 octocat-patch-1@origin b1b3f972
│ │  sentence case
│ ◌  (elided revisions)
├─╯
│ ◉  kowxouwz octocat@nowhere.com 2014-06-10 15:22:26.000 -07:00 test@origin b3cbd5bb
│ │  Create CONTRIBUTING.md
│ ◌  (elided revisions)
├─╯
◉  zzzzzzzz root() 00000000
```

I would prefer this (cleaner and less noisy, illustrated here with VERTICAL FOUR DOTS instead of DOTTED CIRCLE):
```
$ jj log -r 'z|heads(::)'
@  wzkrkuur (no email set) 2024-03-08 11:32:28.000 +02:00 b6b41359
│  (empty) (no description set)
⁞
│ ◉  tpstlust support+octocat@github.com 2018-05-10 12:55:19.000 -05:00 octocat-patch-1@origin b1b3f972
│ │  sentence case
│ ⁞
├─╯
│ ◉  kowxouwz octocat@nowhere.com 2014-06-10 15:22:26.000 -07:00 test@origin b3cbd5bb
│ │  Create CONTRIBUTING.md
│ ⁞
├─╯
◉  zzzzzzzz root() 00000000
```

The effect is much more pronounced when the template renders only one line per commit.

**Reply by martinvonz** [reply]
I included "(elided revisions)" to help new users understand what the node represents. Even if they don't understand what exactly it means, it gives them a phrase to search docs/forums/web for.

**Reply by algmyr** [reply]
FWIW making the node characters configurable would be a good start for the first part, I might look into actually implementing that. It was mentioned recently in https://github.com/martinvonz/jj/issues/3111#issuecomment-1957855885

**Reply by martinvonz** [reply]
I created #3260 for that

**Reply by martinvonz** [reply]
> The current choice of Unicode character (DOTTED CIRCLE) renders as double the width when followed by a space in my terminal (Kitty on Ubuntu Jammy 22.04):

Is that different from the non-ellipsis nodes? That is, does `printf '\u25c9 \n'` render as single width? That's really surprising in that case. I wonder what the reasoning would be.

**Reply by amiryal** [reply]
Yes, FISHEYE always renders as single width regardless of what comes after it.

I did some searching now, and came up with [this FAQ entry for Kitty](https://sw.kovidgoyal.net/kitty/faq/#some-special-symbols-are-rendered-small-truncated-in-kitty). Apparently, some characters are given a special treatment, and my guess is in older versions DOTTED CIRCLE used to be one of them. :shrug:

**Reply by martinvonz** [reply]
Can you confirm that the harpoons symbol works well in your terminal? If it does, I think we should just get #3124 merged. @thoughtpolice FYI.

**Reply by ilyagr** [reply]
> Yes, FISHEYE always renders as single width regardless of what comes after it.
> 
> I did some searching now, and came up with [this FAQ entry for Kitty](https://sw.kovidgoyal.net/kitty/faq/#some-special-symbols-are-rendered-small-truncated-in-kitty). Apparently, some characters are given a special treatment, and my guess is in older versions DOTTED CIRCLE used to be one of them. 🤷

I don't get this problem on Kitty 0.32.1. Dotted circle is single-character width on Mac OS with the Fira Code font.

<img width="483" alt="image" src="https://github.com/martinvonz/jj/assets/4123047/23a6c3ad-4bff-46a0-bf02-d6ad0df9fac9">

I like the dotted circle, but I wouldn't mind replacing it if it doesn't render correctly for many people.


**Update:** After looking at https://github.com/martinvonz/jj/pull/3124, that looks fine to me too. I like it less than the dotted circle, but it seems like the circle is problematic in many fonts.

I would also consider `~`.

**Reply by amiryal** [reply]
Ugh, this is so annoying, LEFTWARDS HARPOON OVER RIGHTWARDS HARPOON also renders in two columns when there is a space after it! (kitty 0.21.2)

I think, if users are comfortable with installing Rust 1.76 or prebuilt binaries from Github, then having a recent non-buggy terminal emulator seems like a reasonable demand.

**Reply by benbrittain** [reply]
I have mixed feelings on that statement. For example, I use jj at work and while I'm able to control binaries installed on my VM/dev box/etc, I don't have control over the terminal emulator on my laptop. I don't think that's an uncommon situation

**Reply by martinvonz** [reply]
Maybe it's fine after #3263 has landed?

**Reply by thoughtpolice** [reply]
Well, it's "you need a non-buggy emulator *if* you want elided nodes." With #3263, realistically speaking &mdash; if you can't find _any_ elided node that works well due to your terminal, then my suggestion would be to just turn off elided nodes... There's nothing we can really do at that point.

**Reply by martinvonz** [reply]
There's always `=` as fallback :) It actually looks pretty okay as a harpoon substitute.

**Reply by benbrittain** [reply]
I think this is fine, especially if #3263 lands. I just want to express caution about assuming everyone has an easy time upgrading their software around jj.
